src/uu/dd/src/bufferedoutput.rs

```diff
@@ -0,0 +1,196 @@
+//  * This file is part of the uutils coreutils package.
+//  *
+//  * For the full copyright and license information, please view the LICENSE
+//  * file that was distributed with this source code.
+//
+// spell-checker:ignore wstat towrite cdefg bufferedoutput
+//! Buffer partial output blocks until they are completed.
+//!
+//! Use the [`BufferedOutput`] struct to create a buffered form of the
+//! [`Output`] writer.
+use crate::{Output, WriteStat};
+
+/// Buffer partial output blocks until they are completed.
+///
+/// Complete blocks are written immediately to the inner [`Output`],
+/// but partial blocks are stored in an internal buffer until they are
+/// completed.
+pub(crate) struct BufferedOutput<'a> {
+    /// The unbuffered inner block writer.
+    inner: Output<'a>,
+
+    /// The internal buffer that stores a partial block.
+    ///
+    /// The size of this buffer is always less than the output block
+    /// size (that is, the value of the `obs` command-line option).
+    buf: Vec<u8>,
+}
+
+impl<'a> BufferedOutput<'a> {
+    /// Add partial block buffering to the given block writer.
+    ///
+    /// The internal buffer size is at most the value of `obs` as
+    /// defined in `inner`.
+    pub(crate) fn new(inner: Output<'a>) -> Self {
+        let obs = inner.settings.obs;
+        Self {
+            inner,
+            buf: Vec::with_capacity(obs),
+        }
+    }
+
+    /// Flush the partial block stored in the internal buffer.
+    pub(crate) fn flush(&mut self) -> std::io::Result<WriteStat> {
+        let wstat = self.inner.write_blocks(&self.buf)?;
+        let n = wstat.bytes_total;
+        for _ in 0..n {
+            self.buf.remove(0);
+        }
+        Ok(wstat)
+    }
+
+    /// Get a mutable reference to the inner block writer.
+    pub(crate) fn get_mut(&mut self) -> &mut Output<'a> {
+        &mut self.inner
+    }
+
+    /// Synchronize the inner block writer.
+    pub(crate) fn sync(&mut self) -> std::io::Result<()> {
+        self.inner.sync()
+    }
+
+    /// Write the given bytes one block at a time.
+    ///
+    /// Only complete blocks will be written. Partial blocks will be
+    /// buffered until enough bytes have been provided to complete a
+    /// block. The returned [`WriteStat`] object will include the
+    /// number of blocks written during execution of this function.
+    pub(crate) fn write_blocks(&mut self, buf: &[u8]) -> std::io::Result<WriteStat> {
```

Comment:
This function also feels like it would be quite slow? I think `concat` will clone the contents and pushing the bytes one by one might be slow, though of course it's hard to know without profiling. Maybe something like this is possible:
```rust
let n = self.buf.len() + buf.len();
let remainder = n % self.inner.settings.obs;
let (to_write, to_buffer) = buf.split_at(buf.len() - remainder);

self.buf.extend_from_slice(to_write);
let wstat = self.inner.write_blocks(&self.buf)?;
self.buf.clear();

self.buf.extend_from_slice(to_buffer);

Ok(wstat)
```

Reply:
Great suggestion, I'll give that a shot and report back.